### PR TITLE
feat(download): Implement download functionality

### PR DESCRIPTION
**Pull request by SaeidRp**
## 🚀 Changes proposed by this PR

This pull request implements the `Download` feature, adds example files demonstrating its usage, and updates the README to include these examples.

### 🧰 Type of change

- [x] New feature (non-breaking change which adds functionality).

### 🤝 Requirements

- [x] ✍ I have read and agree to this projects [Code of Conduct](../../blob/master/.github/CODE_OF_CONDUCT.md).
- [x] ✍ I have read and agree to this projects [Contribution Guidelines](../../blob/master/.github/CONTRIBUTING.md).
- [x] ✍ I have read and agree to the [Developer Certificate of Origin](https://developercertificate.org/).
- [x] 🔎 I have performed a self-review of my own changes.
- [x] 🎨 My changes follow the style guidelines of this project.

- [x] 💬 My changes as properly commented, primarily for hard-to-understand areas.
- [x] 📝 I have made corresponding changes to the documentation.

### CODE REVIEW

**Comment by SaeidRp** [comment]
Hi @lrstanley,

I hope you're doing well!

I wanted to follow up on this pull request. Is there any feedback or suggestions you might have? I'm happy to make any necessary changes to get this merged.

**Comment by lrstanley** [comment]
I haven't had a chance to fully review and test the code, however, a few initial thoughts:

1. If at all possible, we should allow other output (e.g. `print-json`), if yt-dlp supports printing both. The idea would be similar to the way we intercept JSON output on a line-by-line basis (only for lines that match JSON), we could update the progress as we see something that matches a valid progress line. We would also skip the logic completely if no progress function is provided.
2. I'd prefer `Download()` instead be turned into a builder function, something like `SetProgressFn(<function-which-receives-progress>)`, where `Run()` still invokes the download, that way there aren't two different ways of invoking yt-dlp. It would also be good to ensure that the documentation for the `SetProgressFn()`, if it **has to set or disable some specific flags**, that those are explicitly called out in the comment.

**Comment by SaeidRp** [comment]
I agree, using a Builder function would make this much cleaner.

Regarding the simultaneous printing of JSON and progress, it seems yt-dlp currently doesn't support this. Enabling any JSON printing options prioritizes JSON output and skips the actual download. However, it only makes sense to unset those flags in the Download() method.

I'll look into it and keep you updated.

**Comment by SaeidRp** [comment]
@lrstanley I've finished the new implementation. Here is a summary of the changes:

- The progress callback can be set by invoking the `SetProgressFn` method.
- The code now supports handling both single and multiple download progress.
- It tracks playlist progress by adding `PlaylistCount` and `PlaylistIndex` to the `DownloadProgress` struct.
- It stores the download results.

Please review the changes and let me know if any further modifications are needed.

File: command.go [hunk 1]
```diff
@@ -224,6 +236,18 @@ func (c *Command) runWithResult(cmd *exec.Cmd) (*Result, error) {
 // and returns the results (stdout/stderr, exit code, etc). args should be the
 // URLs that would normally be passed in to yt-dlp.
 func (c *Command) Run(ctx context.Context, args ...string) (*Result, error) {
+	if c.progressFunc != nil {
+		c.Progress()
+		c.ProgressTemplate("dl:{\"id\":\"%(info.id)s\",\"playlist_id\":\"%(info.playlist_id)s\",\"title\":\"%(info.title)s\",\"status\":\"%(progress.status)s\",\"total_bytes\":\"%(progress.total_bytes)s\",\"total_bytes_estimate\":\"%(progress.total_bytes_estimate)s\",\"downloaded_bytes\":\"%(progress.downloaded_bytes)s\",\"speed\":\"%(progress.speed)s\",\"percent_str\":\"%(progress._percent_str)s\",\"playlist_count\":\"%(info.playlist_count)s\",\"playlist_index\":\"%(info.playlist_index)s\"}\n")
```
Comment (lrstanley) [inline command.go:241]: Can we be confident that some title might not have one of:

- newlines (how would this be handled?)
- quote symbols and/or something that JSON doesn't directly support?

I am not super familiar with yt-dlp and it's escaping rules for `%(field)s` type fields.

To further clarify on my previous comments, I don't necessarily think the format has to be JSON, just handled in the same way that timestamp-based logging happens. I'm not sure if JSON here is better or worse (assuming there is potential for weird characters to cause JSON to break).

File: download.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+	StatusDownloading DownloadStatus = "downloading"
+	// StatusPostProcessing represents the download status when the download is post-processing.
+	StatusPostProcessing DownloadStatus = "post_processing"
+	// StatusFinished represents the download status when the download is finished.
+	StatusFinished DownloadStatus = "finished"
+)
+
+// DownloadProgress represents the progress of a download.
+type DownloadProgress struct {
+	ID            string
```
Comment (lrstanley) [inline download.go:28]: Can we add documentation to these fields? E.g. where they are populated from, what `Total` means, etc. If possible, include the unit of measurement in the field name (e.g. is `Total` the `TotalBytes`?)

File: results.go [hunk 3]
```diff
@@ -133,6 +136,10 @@ type timestampWriter struct {
 	buf            bytes.Buffer
 	lastWriteStart time.Time
 	results        []*ResultLog
+	downloads      map[string]*DownloadProgress
+
+	progress     DownloadProgress
```
Comment (lrstanley) [inline results.go:141]: Looks like this field is unused.

**Review by lrstanley — CHANGES_REQUESTED** [review]

File: command.go (hunk 1 above)
Comment (SaeidRp) [reply command.go:241]: You are correct! Fortunately, yt-dlp supports printing fields with escaped double quotes using the %()j formatting option. I will update the template to utilize this option for proper double-quote handling.

**Comment by SaeidRp** [comment]
regarding concerns around the consistency and safety of value parsing in our current implementation. I’ve implemented a new template encoding system that ensures safe parsing and formatting.

Key improvements:

- **Template Encoding (Marshal/Unmarshal):** Provides a standardized and reliable way to handle ytdlp output, preventing potential parsing errors and improving consistency.
- **Template Formatter:** Allows for flexible modification of output values before casting, enabling greater control over data manipulation.
- **Extensive Testing:** Given the critical role of template parsing, I've thoroughly tested the new code to ensure its reliability.

**Comment by SaeidRp** [comment]
@lrstanley, Just checking in on this PR. I submitted it last week and would love to get your feedback whenever you have a chance.

**Comment by lrstanley** [comment]
I am not confident at this point if the solution you've proposed is in fact better. From initial glance, feels much more complicated than required. I simply haven't had the time in the past couple of weeks to look into it further.

**Comment by SaeidRp** [comment]
No worries about the delay – I get that things can pile up and it takes time to properly review changes.
I've decided to forge ahead on my fork for now. This lets me implement and test changes at the pace I need.

Your repo has been a solid starting point. I appreciate the work you've put into it.

**Comment by lrstanley** [comment]
@SaeidRp @eos175 -- unless I'm missing something, which includes the implementation in this PR and the one you have @eos175 in your repo, yt-dlp supports outputting JSON and progress, including still doing the download?

Was testing this with the latest release, and it:
1. downloads the file
2. outputs the necessary JSON
3. outputs progress in the format that I asked.

Am I missing something as far as that limitation?

Example output:

```console
$ yt-dlp \
  --print-json \
  --format-sort res,ext:mp4:m4a \
  --recode-video mp4 \
  --no-playlist \
  --continue \
  --output "%(extractor)s - %(title)s.%(ext)s" \
  --newline \
  --progress \
  --progress-template "dl:%(progress.total_bytes)s,%(progress.downloaded_bytes)s" \
  https://cdn.liam.sh/github/go-ytdlp/sample-1.mp4
{"id": "sample-1", "title": "sample-1", "timestamp": 1693885266.0, "direct": true, "formats": [{"format_id": "mp4", "url": "https://cdn.liam.sh/github/go-ytdlp/sample-1.mp4", "ext": "mp4", "vcodec": null, "protocol": "https", "resolution": null, "dynamic_range": "SDR", "aspect_ratio": null, "filesize_approx": null, "http_headers": {"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.107 Safari/537.36", "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8", "Accept-Language": "en-us,en;q=0.5", "Sec-Fetch-Mode": "navigate"}, "video_ext": "mp4", "audio_ext": "none", "vbr": null, "abr": null, "tbr": null, "format": "mp4 - unknown"}], "subtitles": {}, "http_headers": {"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.107 Safari/537.36", "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8", "Accept-Language": "en-us,en;q=0.5", "Sec-Fetch-Mode": "navigate"}, "hls_aes": null, "webpage_url": "https://cdn.liam.sh/github/go-ytdlp/sample-1.mp4", "original_url": "https://cdn.liam.sh/github/go-ytdlp/sample-1.mp4", "webpage_url_basename": "sample-1.mp4", "webpage_url_domain": "cdn.liam.sh", "extractor": "generic", "extractor_key": "Generic", "playlist": null, "playlist_index": null, "display_id": "sample-1", "fulltitle": "sample-1", "upload_date": "20230905", "release_year": null, "requested_subtitles": null, "_has_drm": null, "epoch": 1721273689, "format_id": "mp4", "url": "https://cdn.liam.sh/github/go-ytdlp/sample-1.mp4", "ext": "mp4", "vcodec": null, "protocol": "https", "resolution": null, "dynamic_range": "SDR", "aspect_ratio": null, "filesize_approx": null, "video_ext": "mp4", "audio_ext": "none", "vbr": null, "abr": null, "tbr": null, "format": "mp4 - unknown", "_filename": "generic - sample-1.mp4", "filename": "generic - sample-1.mp4", "_type": "video", "_version": {"version": "2024.07.16", "current_git_head": null, "release_git_head": null, "repository": "yt-dlp/yt-dlp"}}
dl:4745623,1024
dl:4745623,3072
dl:4745623,7168
dl:4745623,15360
dl:4745623,31744
dl:4745623,64512
dl:4745623,130048
dl:4745623,261120
dl:4745623,523264
dl:4745623,1047552
dl:4745623,2096128
dl:4745623,4193280
dl:4745623,4745623
dl:4745623,4745623
```

aaand:

```console
$ ll *.mp4
-rw-r--r-- 1 liam liam 4.6M Sep  4  2023 'generic - sample-1.mp4'
```

**Comment by eos175** [comment]
I confirm that with the latest version `2024.07.16`, there is progress and JSON in my repo with:

```
func (c *Command) SetProgress(delta time.Duration, fun CallbackProgress) *Command {
    c.Progress().
        ProgressTemplate(`dl:%(progress.total_bytes)s,%(progress.downloaded_bytes)s`).
        ProgressDelta(float64(delta.Seconds())).
        Newline()
    c.mu.Lock()
    c.fun = fun
    c.mu.Unlock()
    return c
}
```

**Comment by SaeidRp** [comment]
Using the `print-json` flag will not interrupt the download (if I recall correctly), but `dump-json` does.

**Pull request closed without merging**